### ISSUE TITLE
Issue with NaNs in CTC loss

### ISSUE THREAD

**Issue by rishabh135** [espnet/espnet#4634]
**Describe the bug**
I am trying to train ESPNet on librispeech for ASR task with modified learning rate for encoder decoder and facing this issue of 

```All samples in this mini-batch got nan grad. Returning nan value instead of CTC loss
WARNING:root:The grad norm is nan. Skipping updating the model.
```

The modified codebase is as follows :
```
    @classmethod
    def build_optimizers(
        cls,
        args: argparse.Namespace,
        model: torch.nn.Module,
    ) -> List[torch.optim.Optimizer]:
        
        if cls.num_optimizers != 1:
            raise RuntimeError(
                "build_optimizers() must be overridden if num_optimizers != 1"
            )

        optim_class = optim_classes.get(args.optim)
        if ( args.adv_flag):
            
            if(args.ngpu > 1):        
                param_grp = [
                    {'params': model.module.encoder.parameters(), 'lr': args.asr_lr},
                    {'params': model.module.decoder.parameters(), 'lr': args.asr_lr},
                    {'params': model.module.ctc.parameters(), 'lr': args.asr_lr},
                    {'params': model.module.adversarial_branch.parameters(), 'lr': args.adv_lr}
                ]            
            else:
                param_grp = [
                    {'params': model.encoder.parameters(), 'lr': args.asr_lr},
                    {'params': model.decoder.parameters(), 'lr': args.asr_lr},
                    {'params': model.ctc.parameters(), 'lr': args.asr_lr},
                    {'params': model.adversarial_branch.parameters(), 'lr': args.adv_lr}]
            
            optimi = optim_class(param_grp)
        else:
            optimi = optim_class(model.parameters(), **args.optim_conf)

        optimizers = [optimi]
        return optimizers
    ```

The bug seems to arrrive randomly on the 4th or 5th epoch while training and leads to complete garbage of results

**Basic environments:**
 - OS information: [e.g., Linux 4.9.0-11-amd64 #1 SMP Debian 4.9.189-3+deb9u2 (2019-11-11) x86_64]
 - python version: [e.g. 3.7.3 (default, Mar 27 2019, 22:11:17) [GCC 7.3.0]]
 - espnet version: [e.g. espnet 0.8.0]
 - Git hash [e.g. b88e89fc7246fed4c2842b55baba884fe1b4ecc2]
   - Commit date [e.g. Tue Sep 1 09:32:54 2020 -0400]
 - pytorch version [e.g. pytorch 1.4.0]
 
You can obtain them by the following command
```
cd <espnet-root>/tools
. ./activate_python.sh

echo "- OS information: `uname -mrsv`"
python3 << EOF
import sys, espnet, torch
pyversion = sys.version.replace('\n', ' ')
print(f"""- python version: \`{pyversion}\`
- espnet version: \`espnet {espnet.__version__}\`
- pytorch version: \`pytorch {torch.__version__}\`""")
EOF
cat << EOF
- Git hash: \`$(git rev-parse HEAD)\`
  - Commit date: \`$(git log -1 --format='%cd')\`
EOF
```

**Environments from `torch.utils.collect_env`:**
e.g.,
```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: CentOS Linux release 7.5.1804 (Core)
GCC version: (GCC) 7.2.0
CMake version: version 2.8.12.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: TITAN RTX
GPU 1: TITAN RTX
GPU 2: TITAN RTX
GPU 3: TITAN RTX

Nvidia driver version: 440.33.01
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] pytorch-ranger==0.1.1
[pip3] pytorch-wpe==0.0.0
[pip3] torch==1.4.0
[pip3] torch-complex==0.1.1
[pip3] torch-optimizer==0.0.1a14
[pip3] torchaudio==0.4.0
[pip3] warprnnt-pytorch==0.1
[conda] blas                      1.0                         mkl
[conda] mkl                       2020.1                      217
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.1.0            py37h23d657b_0
[conda] mkl_random                1.1.1            py37h0573a6f_0
[conda] pytorch                   1.4.0           py3.7_cuda10.0.130_cudnn7.6.3_0    pytorch
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] pytorch-wpe               0.0.0                    pypi_0    pypi
[conda] torch-complex             0.1.1                    pypi_0    pypi
[conda] torch-optimizer           0.0.1a14                 pypi_0    pypi
[conda] torchaudio                0.4.0                    pypi_0    pypi
[conda] warprnnt-pytorch          0.1                      pypi_0    pypi
```
You can obtain them by the following command

```
cd <espnet-root>/tools
. ./activate_python.sh
python3 -m torch.utils.collect_env
```

**Task information:**
 - Task: [e.g., ASR, TTS, ST, ENH]
 - Recipe: [e.g. librispeech]
 - ESPnet1 or ESPnet2

**To Reproduce**
1. move to a recipe directory, e.g., `cd egs/librispeech/asr1`
2. execute `run.sh` with specific arguments, e.g., `run.sh --stage 3 --ngp 1`
3. specify the error log, e.g., `exp/xxx/yyy.log`

**Error logs**
Paste the error logs. If applicable, add screenshots to help explain your problem.

**Comment by rishabh135** [comment]
Thanks @sw005320 Shinji-san for your quick response, it's not a rare occurrence, infact [if you can take a look at the training logs](https://justpaste.it/4e7qg), it starts occurring just randomly out of blue. 

What do you mean by more tuning ? Are you referring to tuning learning rate parameters ? If so, how it can be that when I train with the same LR without this split across params in the optimizer, it works, without a hitch.  

Can you please share some more details about this ?

Thanks again for your help!

**Comment by sw005320** [comment]
This happens when the encoder output is shorter than the label length, or you just set the wrong optimizer.
Please check both cases.
The second case is specific to your problem, and we cannot help so much.

**Comment by rishabh135** [comment]
Thanks @sw005320, I figured out the error. You are right, the issue occurs due to exploding CTC loss.

 I do have a pretty basic question, which hopefully you'll forgive me for asking, I have a network, where I have the optimizer as illustrated above and depending upon configuration ( i.e. when I want a certain subsection of the network frozen), I set up certain layers of the model as frozen, in that scenario why is that my CTC loss is updated ? How can I freeze the CTC loss as well during that time and  later on in the training loop when I unfreeze the decoder layers, retrieve it from the frozen standpoint ? 
 
 Is there a way for me to ensure since there is only one optimizer to have certain loss as frozen during certain epochs ?
 
 Thanks and once again, grateful for your help and guidance with regards to this error!

**Comment by rishabh135** [comment]
More concretely, I have the following [CTC loss graph](https://wandb.ai/inria_marc/nancy_v2_adv_sep_16_freeze_ctcASRTask/reports/train-train_loss_ctc_epoch-22-09-16-12-35-26---VmlldzoyNjQ0MzY3?accessToken=i9oapb5yweinf6wv0zndp22keyb41jud78muvzgwo3xjvjtrj1hm52hkcgyq2988), I am confused why all of a sudden between 20th and 42th epoch, there is such a change in CTC loss, why does it explode all of a sudden ? I am not making any changes in the computation graph or applying any different variation during this time? 
What causes this sudden explosion ? 

It would be great if you can help me out! Thanks so very much!

**Comment by sw005320** [comment]
I don't think this is sudden.
There is some small corruption already around 10 epochs, and the loss gradually increases (training is not regular). 
Due to this wrong training behavior, I feel this is very natural that further "sudden" corruption happen later.
This is not due to the architecture (not a computational graph). This comes from an issue of wrong optimizer setup.

It would be great if you understand this kind of training behavior and tune optimization hyper-parameters by yourself according to your set up.

**Comment by rishabh135** [comment]
Thanks @sw005320 Watanabe-san, I followed your advice and looked into tuning the hyperparameters by myself. As you've worked so much on pytorch for this system, can I please ask you something that's bugging me? 

When I freeze some layers of the network using 
```
def freeze_encoder(self):
    for param in self.encoder.parameters():
        param.requires_grad = False
        if param.grad is not None:
            param.grad.zero_()

 ```
Why is that my attention_loss still varies (updates)  even though [I am freezing the  encoder on the 11th epoch](https://wandb.ai/inria_marc/nancy_v4_oct_6_standard_170ASRTask/reports/train-train_loss_att_epoch-22-10-06-13-12-02---VmlldzoyNzQ5ODUz?accessToken=8jwgleh49icvnadalor231jyjcoqfqhkptr5bpvnpvxhfjyr1v3zm0yoknukemaf) and hence the attention loss calculated across must remain constant across the 10th and 11th epoch  ?

Is that again some sort of dynamic data corruption that is happening, or is it due to some stochastically in the standard conformer encoder? I have also set the `cudnn_deterministic: true `   and I have a constant seed `seed: 2022` during my training.

This is something that I've faced for a long time and would help me enormously if you can help me understand why is it happening.

**Comment by sw005320** [comment]
The attention loss is computed from the encoder and decoder networks.
So, it would be changed if you only fix the encoder network.

**Comment by sw005320** [comment]
> Thanks, @sw005320, though I was freezing both the encoder and decoder networks, it was still not working, but I was able to finally resolve the issue using `optimizer.zero_grad(set_to_none=True)` in trainer.py and setting the grads as explicitly none :
> 
> ```
> def freeze_encoder(self):
>         if not self.encoder_frozen_flag:
>             for param in self.encoder.parameters():
>                 param.requires_grad = False
>                 param.grad = None
>                 # if param.grad is not None:
>                 #     param.grad.zero_()
>             for param in self.decoder.parameters():
>                 param.requires_grad = False
>                 param.grad = None
>                 # if param.grad is not None:
>                 #     param.grad.zero_()
>                 
>             for param in self.ctc.ctc_lo.parameters():
>                 param.requires_grad = False
>                 param.grad = None
>                 
> ```
> 
> If folks want to do end to end training with different phases, this should help them as well!

Thanks for your TIPS.
@DanBerrebbi and @simpleoier, did you know this?
If this is true, many of our transfer learning methods could have an issue.
Should we fix it?

> Can I also please ask, as a sidenote, what sort of decoder should I use for feats reconstruction ? i.e. with which I am able to recover the correct Mel spectra features ?
> 
> Thanks, and this issue is closed now from my end!

This depends on what you want to use.
I would probably use some simple architecture (e.g., transformer, LSTM, etc.).
My concern is the upsampling part, and I'm not sure that a transformer with upsampling (whether reshaping or positional encoding with the original feature length) can well recover the length.

**Comment by DanBerrebbi** [comment]
In my experience and if I understand well your (solved) issue here, I indeed encountered some Nan in CTC loss in some transfer learning setups. For my case it was mostly due to too short input because of subsampling which you can solve through config file by specifying (for transformer for instance) the subsampling factor by mentioning the stride of convolution that you want as done [here](https://github.com/espnet/espnet/blob/master/egs2/accented_french_openslr57/asr1/conf/tuning/train_asr_transformer.yaml#L5 ). 
This did not affect performances for my case so I was happy enough with that solution even though others may be more suitable.

**Comment by simpleoier** [comment]
@sw005320 I think this is interesting. What we usually did is just setting non-updated parameters to `param.requires_grad = False`. We didn't do this further step `param.grad = None`. We need to do further tests on this.
@rishabh135 Thanks for this interesting finding. But I may missed something in the discussion. Can you help me to understand what `optimizer.zero_grad(set_to_none=True)` is used for? In my understanding, it sets all params of the optimizer to be `param.grad = None`. Did you use it and the following code together?

> ```
>   def freeze_encoder(self):
>         if not self.encoder_frozen_flag:
>             for param in self.encoder.parameters():
>                 param.requires_grad = False
>                 param.grad = None
>                 # if param.grad is not None:
>                 #     param.grad.zero_()
>             for param in self.decoder.parameters():
>                 param.requires_grad = False
>                 param.grad = None
>                 # if param.grad is not None:
>                 #     param.grad.zero_()
>                 
>             for param in self.ctc.ctc_lo.parameters():
>                 param.requires_grad = False
>                 param.grad = None
> ```

**Comment by sw005320** [comment]
Yes, frankly, I'm also confused, and it would be great if @rishabh135 clarifies it.
